### ISSUE TITLE
Implement Ruby style layout for Records / Structs

### ISSUE THREAD

**Issue by plicease** [plicease/FFI-Platypus#21]
There are (today) a few ways of dealing with C structs (or their equivalents in other languages):

 - Perl's pack/unpack
 - Convert::Binary::C
 - Inline::Struct
 - bundle your extension with a ffi directory or use FFI::TinyCC to write accessors in 

Some examples of these would be great.

Also a better solution would be great too.  Perhaps something like ruby ffi's layout system could be borrowed/stolen.

**Comment by plicease** [comment]
http://blogs.perl.org/users/graham_ollis/2015/01/practical-ffi-with-platypus.html#comment-1531389

**Comment by plicease** [comment]
Supporting a ruby style layout interface shouldn't be too hard.  I have implemented alignof:

adb99fb3b57a53a1af235269678c33d7970e4bc0

**Comment by plicease** [comment]
This implement record members for uint8 and uint32.  Banging out the rest of the integers, floats and opaque types should be pretty easy.

46c1dc2f2ee77acd88957a6ee4a734bb595dccc3

Interesting to consider what to do about strings.  I'd like to hang them off the object some how and free them automatically when they fall out of scope.  Ought to be doable, maybe with magic or something.

Failing that a `DESTROY` and some inside outness of objectivity ought to be able to do it.

**Comment by plicease** [comment]
This feature should block the release of 0.21 unless something more pressing comes up.

**Comment by plicease** [comment]
alignment detection seems to break on gcc (I usually use clang)

https://travis-ci.org/plicease/FFI-Platypus/builds/49675771

**Comment by plicease** [comment]
I am definitely going to implement arrays as record members to handle things like this:

```
struct {
  int int_array[200];
  char string[200];
};
```

That is easy.

But how much hand holding should we do for things like:

```
struct {
  int *int_pointer;
  char *string;
}
```

Approaches:

  - don't support them at all, let the users define them as `opaque` and use their own manipulations
  - allocate when set / free when DESTROY ... challenge here is remembering when WE allocated the memory vs. when C allocated the memory.
  - something else

**Comment by plicease** [comment]
challenge with the allocate when set / free when DESTROY is making sure C didn't update the pointer in the meantime.  Here is another approach:

  - string rw allocate when set / free when DESTROY, does not support C modifications
  - string ro no ability to set / but can read values set by C

**Comment by plicease** [comment]
Another option is to go with this point:

  - don't support them at all

And handle any required use cases with a wrapper:

```
record_layout(
  opaque => field_name => sub {
    # a string
    my($self, $value) = @_;
    if(defined $value)
    {
      my $ptr = strdup($value);
      $self->_field_name($ptr);  # alternative pass in the $inner as an argument
      return $value;
    }
    return $ffi->cast(opaque => string => $self->_field_name;
  },
);

sub DESTROY
{
  my($self) = @_;
  if(defined $self->_field_name)
  {
    free($self->_field_name);
  }
}
```

**Comment by plicease** [comment]
how to handle embedded records and unions:

```
struct foo1 {
  char mess_up_alignment;
  struct bar {// alignof bar == alignof long
    short baz1;
    long baz2;
  };
};

struct foo2 {
  char mess_up_alignment;
  union bar { // alignof bar == alignof long
    int baz1;
    char baz2;
    long baz3;
  };
};
```

you should be able to do this:

```
my $value = $foo1->bar->baz1
```

but $foo1 is a packed string.  so to set:

```
  # this is awkward
  my $bar = $foo1->bar;
  $bar->baz1($value);
  $foo1->bar($bar);
```

**Comment by plicease** [comment]
This is implemented in 0.21.  Still nice to have are nested records, and unions.  Most other things can be implemented through the use of accessor wrappers around opaque types.
